### DISCUSSION TITLE
Problems with PCA9685

### DISCUSSION THREAD

**Discussion by Basti149** [arendst/Tasmota#22422] [General]
I've been trying to put my PCA9685 into operation with Tasmota for days. But unfortunately it doesn't work. 
After he always showed the INA219. I deactivated driver 14, which was never plugged in, see picture. 
But the PCA9685 still doesn't work. 
As you can see in the picture it is recognized with 0x40 (0x20 is an MCP23017 on the same bus).
![grafik](https://github.com/user-attachments/assets/04a44371-e63e-460b-8324-56107fdebeb6)

I compiled the following lines but it doesn't work.
![grafik](https://github.com/user-attachments/assets/4f1acefc-1178-464c-aa27-16356182edb4)

Do I need to enter anything else into the console? Except I2C driver1 1

**Comment by barbudor** [comment]
Do you have any log at start that would confirm the PCA9685 is actually detected as a PC9685 by the driver ?

Set weblog 4, restart and provide the logs here using proper copy/paste with triple-backtics, not screen shot

**Reply by Basti149** [reply]
EDIT


00:00:00.001 HDW: ESP32-C2 v1.0 
00:00:00.017 UFS: FlashFS mounted with 304 kB free
00:00:00.027 CFG: Loaded from File, Count 98
00:00:00.035 QPC: Count 1
00:00:00.036 CFG: CR 429/699, Busy 0
00:00:00.042 TFS: File 'mcp23x.dat' not found
00:00:00.042 MCP: Invalid template
00:00:00.042 TYA: Active=0
00:00:00.046 ROT: Mode 1
00:00:00.053 BRY: GC from 3898 to 3698 bytes, objects freed 4/17 (in 0 ms) - slots from 33/61 to 25/61
00:00:00.070 CFG: No '*.autoconf' file found
00:00:00.074 BRY: GC from 5807 to 4260 bytes, objects freed 8/35 (in 1 ms) - slots from 54/76 to 35/76
00:00:00.074 BRY: Berry initialized, RAM used 4260 bytes
00:00:00.080 BRY: No 'preinit.be'
00:00:00.082 SRC: Restart
00:00:00.084 Project tasmota - Tasmota Version 14.3.0(tasmota32)-3_1_0(2024-11-03T19:36:04)
00:00:00.103 DSB: Sensors found 1
00:00:00.115 SHT: Sensor did not ACK command
00:00:00.147 I2C: MCP23017 found at 0x20
00:00:00.446 BRY: No 'autoexec.be'
00:00:00.533 WIF: Checking connection...
00:00:00.533 WIF: Attempting connection...
00:00:02.001 WIF: Connecting to AP1 SmartHome System Mira Channel 7 BSSId 7A:8C:B5:AE:4D:10 in mode HT40 as tasmota-0F7A0C-6668...
00:00:02.139 WIF: IPv4 192.168.1.162, mask 255.255.255.0, gateway 192.168.1.1
00:00:03.698 WIF: IPv6 Local fe80::3698:7aff:fe0f:7a0c%st1
00:00:04.111 WIF: Checking connection...
00:00:04.111 WIF: Connected
00:00:04.317 HTP: Web server active on tasmota-0F7A0C-6668 with IP address 192.168.1.162
00:00:04.463 HTP: Main Menu
00:00:04.653 HTP: Main Menu
00:00:04.984 NTP: Sync time...
00:00:05.986 WIF: DNS failed for 2.pool.ntp.org after 1001 ms
00:00:06.002 NTP: Sync time...
00:00:07.003 WIF: DNS failed for 2.europe.pool.ntp.org after 1000 ms
00:00:07.038 RSL: INFO1 = {"Info1":{"Module":"ESP32C2","Version":"14.3.0(tasmota32)","FallbackTopic":"cmnd/DVES_0F7A0C_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:07.051 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-0F7A0C-6668","IPAddress":"192.168.1.162","IP6Global":"","IP6Local":"fe80::3698:7aff:fe0f:7a0c%st1"}}
00:00:07.064 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":46}}
00:00:07.076 RSL: RESULT = {"POWER1":"OFF"}
00:00:07.078 RSL: POWER1 = OFF
00:00:07.079 RSL: RESULT = {"POWER2":"OFF"}
00:00:07.080 RSL: POWER2 = OFF
00:00:07.081 RSL: RESULT = {"POWER3":"OFF"}
00:00:07.092 RSL: POWER3 = OFF
00:00:07.093 RSL: RESULT = {"POWER4":"OFF"}
00:00:07.094 RSL: POWER4 = OFF
00:00:07.095 RSL: RESULT = {"POWER5":"OFF"}
00:00:07.106 RSL: POWER5 = OFF
00:00:07.108 RSL: RESULT = {"POWER6":"OFF"}
00:00:07.108 RSL: POWER6 = OFF
00:00:07.110 RSL: RESULT = {"POWER7":"OFF"}
00:00:07.111 RSL: POWER7 = OFF
00:00:07.122 RSL: RESULT = {"POWER8":"OFF"}
00:00:07.123 RSL: POWER8 = OFF
00:00:07.124 RSL: RESULT = {"POWER9":"OFF"}
00:00:07.125 RSL: POWER9 = OFF
00:00:07.136 RSL: RESULT = {"POWER10":"OFF"}
00:00:07.138 RSL: POWER10 = OFF
00:00:07.139 RSL: RESULT = {"POWER11":"OFF"}
00:00:07.140 RSL: POWER11 = OFF
00:00:07.141 RSL: RESULT = {"POWER12":"OFF"}
00:00:07.152 RSL: POWER12 = OFF
00:00:07.154 RSL: RESULT = {"POWER13":"OFF"}
00:00:07.155 RSL: POWER13 = OFF
00:00:07.156 RSL: RESULT = {"POWER14":"OFF"}
00:00:07.168 RSL: POWER14 = OFF
00:00:07.169 RSL: RESULT = {"POWER15":"OFF"}
00:00:07.170 RSL: POWER15 = OFF
00:00:07.171 RSL: RESULT = {"POWER16":"OFF"}
00:00:07.172 RSL: POWER16 = OFF
00:00:07.187 TFS: File 'autoexec.bat' not found
00:00:07.190 NTP: Sync time...
00:00:08.193 WIF: DNS failed for 2.nl.pool.ntp.org after 1001 ms
00:00:08.208 HTP: Tools
00:00:08.278 QPC: Reset
00:00:08.279 NTP: Sync time...
00:00:09.281 NTP: No reply from 192.168.1.1
00:00:09.305 NTP: Sync time...
00:00:10.306 WIF: DNS failed for 1.pool.ntp.org after 1001 ms
00:00:10.354 APP: Boot Count 46
00:00:10.355 NTP: Sync time...
00:00:10.356 WIF: DNS failed for 2.pool.ntp.org after 0 ms
00:00:10.427 HTP: Console
00:00:10.974 BRY: GC from 6553 to 4588 bytes, objects freed 17/45 (in 1 ms) - slots from 75/91 to 49/91
00:00:10.980 RSL: STATE = {"Time":"1970-01-01T00:00:10","Uptime":"0T00:00:11","UptimeSec":11,"Heap":65,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":304,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":45},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","POWER5":"OFF","POWER6":"OFF","POWER7":"OFF","POWER8":"OFF","POWER9":"OFF","POWER10":"OFF","POWER11":"OFF","POWER12":"OFF","POWER13":"OFF","POWER14":"OFF","POWER15":"OFF","POWER16":"OFF","Wifi":{"AP":1,"SSId":"SmartHome System Mira","BSSId":"7A:8C:B5:AE:4D:10","Channel":7,"Mode":"HT40","RSSI":48,"Signal":-76,"LinkCount":1,"Downtime":"0T00:00:04"}}
00:00:11.053 RSL: SENSOR = {"Time":"1970-01-01T00:00:11","DS18B20":{"Id":"3C01B556A4AC","Temperature":23.3},"MCP230XX":{"D0":1,"D1":1,"D2":1,"D3":1,"D4":1,"D5":1,"D6":1,"D7":1,"D8":1,"D9":1,"D10":1,"D11":1,"D12":1,"D13":1,"D14":1,"D15":1,"OUT":{"OUT_D0":"OFF","OUT_D1":"OFF","OUT_D2":"OFF","OUT_D3":"OFF","OUT_D4":"OFF","OUT_D5":"OFF","OUT_D6":"OFF","OUT_D7":"OFF","OUT_D8":"OFF","OUT_D9":"OFF","OUT_D10":"OFF","OUT_D11":"OFF","OUT_D12":"OFF","OUT_D13":"OFF","OUT_D14":"OFF","OUT_D15":"OFF"}},"ESP32":{"Temperature":22.0},"TempUnit":"C"}
00:00:11.098 NTP: Sync time...
00:00:11.099 WIF: DNS failed for 2.europe.pool.ntp.org after 0 ms
00:00:11.987 NTP: Sync time...
00:00:12.988 WIF: DNS failed for 2.nl.pool.ntp.org after 1000 ms
00:00:13.018 NTP: Sync time...
00:00:14.023 NTP: No reply from 192.168.1.1
00:00:14.075 NTP: Sync time...
00:00:15.077 WIF: DNS failed for 1.pool.ntp.org after 1001 ms
00:00:15.104 NTP: Sync time...
00:00:16.106 WIF: DNS failed for 2.pool.ntp.org after 1001 ms
00:00:16.132 NTP: Sync time...
00:00:17.134 WIF: DNS failed for 2.europe.pool.ntp.org after 1001 ms
00:00:17.146 RSL: config = {"ip":"192.168.1.162","dn":"Tasmota","fn":["Tasmota","Tasmota2","Tasmota3","Tasmota4","","","",""],"hn":"tasmota-0F7A0C-6668","mac":"34987A0F7A0C","md":"ESP32C2","ty":0,"if":0,"ofln":"Offline","onln":"Online","state":["OFF","ON","TOGGLE","HOLD"],"sw":"14.3.0","t":"tasmota_0F7A0C","ft":"%prefix%/%topic%/","tp":["cmnd","stat","tele"],"rl":[1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"swc":[-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1],"swn":[null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null,null],"btn":[0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0],"so":{"4":0,"11":0,"13":0,"17":0,"20":0,"30":0,"68":0,"73":0,"82":0,"114":0,"117":0},"lk":0,"lt_st":0,"bat":0,"dslp":0,"sho":[],"sht":[],"ver":1}
00:00:17.174 RSL: sensors = {"sn":{"Time":"1970-01-01T00:00:17","DS18B20":{"Id":"3C01B556A4AC","Temperature":23.3},"MCP230XX":{"D0":1,"D1":1,"D2":1,"D3":1,"D4":1,"D5":1,"D6":1,"D7":1,"D8":1,"D9":1,"D10":1,"D11":1,"D12":1,"D13":1,"D14":1,"D15":1,"OUT":{"OUT_D0":"OFF","OUT_D1":"OFF","OUT_D2":"OFF","OUT_D3":"OFF","OUT_D4":"OFF","OUT_D5":"OFF","OUT_D6":"OFF","OUT_D7":"OFF","OUT_D8":"OFF","OUT_D9":"OFF","OUT_D10":"OFF","OUT_D11":"OFF","OUT_D12":"OFF","OUT_D13":"OFF","OUT_D14":"OFF","OUT_D15":"OFF"}},"ESP32":{"Temperature":21.0},"TempUnit":"C"},"ver":1}
00:00:17.213 NTP: Sync time...
00:00:18.215 WIF: DNS failed for 2.nl.pool.ntp.org after 1000 ms
00:00:18.229 NTP: Sync time...
00:00:19.232 NTP: No reply from 192.168.1.1
00:00:19.263 NTP: Sync time...
00:00:20.265 WIF: DNS failed for 2.pool.ntp.org after 1000 ms
00:00:20.279 NTP: Sync time...
00:00:20.281 WIF: DNS failed for 2.pool.ntp.org after 0 ms
00:00:20.976 NTP: Sync time...
00:00:20.978 WIF: DNS failed for 2.europe.pool.ntp.org after 0 ms
00:00:21.981 NTP: Sync time...
00:00:21.983 WIF: DNS failed for 2.nl.pool.ntp.org after 0 ms
00:00:22.981 NTP: Sync time...
00:00:23.984 NTP: No reply from 192.168.1.1
00:00:24.002 NTP: Sync time...
00:00:25.005 WIF: DNS failed for 2.pool.ntp.org after 1002 ms
00:00:25.043 NTP: Sync time...
00:00:26.044 WIF: DNS failed for 2.pool.ntp.org after 1000 ms
00:00:26.066 NTP: Sync time...
00:00:27.068 WIF: DNS failed for 2.europe.pool.ntp.org after 1001 ms
00:00:27.092 NTP: Sync time...
00:00:28.094 WIF: DNS failed for 2.nl.pool.ntp.org after 1001 ms
00:00:28.112 NTP: Sync time...
00:00:29.115 NTP: No reply from 192.168.1.1
00:00:29.140 NTP: Sync time...
00:00:29.142 WIF: DNS failed for 1.pool.ntp.org after 0 ms
00:00:31.445 BRY: GC from 9509 to 4987 bytes, objects freed 10/45 (in 0 ms) - slots from 49/91 to 49/91
00:00:39.925 WIF: Checking connection...
00:00:39.976 NTP: Sync time...
00:00:40.979 WIF: DNS failed for 2.pool.ntp.org after 1001 ms

**Reply by Basti149** [reply]
It always recognized the INA219, but it was never connected. I have therefore set Driver14 to 0.  It still displayed 0x70, 0,72 before, but they are gone. These addresses never existed but only these two, which are now displayed correctly.
The address 0x20 is the MCP23017
Address 0x40 should be the PCA9685

**Reply by barbudor** [reply]
> It always recognized the INA219, but it was never connected

I don't understand what you mean by this sentence

But from the logs, the PCA9685 is not recognized
If it was recognized you would have a line similar to the one of the MCP:
```
 I2C: PCA9685 found at 0x40
```

So you have something wrong
PCA is i2cdriver 1 so it should be tested first. 
Nevertheless you have too many active drivers, good reason for possible confusion.
So disable all drivers but driver 1 (and driver 22 or 77 for the MCP depending if you are using mode 1 or mode 2)

**Reply by Basti149** [reply]
> I don't understand what you mean by this sentence
> 
> But from the logs, the PCA9685 is not recognized If it was recognized you would have a line similar to the one of the MCP

Yes, I understand that, but the address must be the PCA9685. Nothing else is connected apart from these two.



Ok then I deactivate the driver I have already deactivated the driver 14 because otherwise it showed me the INA219 which was and is never connected.


Brief history The board in conjunction with the ESP went at the beginning of the year, after which I put it in the cupboard as I didn't need it yet. And about last week I took it out and wanted to try it again and suddenly the GUI showed IN219 with voltage, current and power, which made me wonder.

**Reply by barbudor** [reply]
> Ok then I deactivate the driver 

Not sure what you understood
I wrote to disable all drivers but driver 1
That means disable all other drivers except driver 1 which has to remains enabled

**Comment by Basti149** [comment]
Yes, I understood that. But I have the question of why it shows the Ina219 even though I haven't completed it. I marked the INA with # in visual studio but it still showed it to me in the Gui.
Only when I deactivated it using I2CDriver 14 0. I think he was gone in the Gui but he still thinks that's why the PCA9685 doesn't come.

1. Now the question arises whether I didn't compile it correctly or whether it didn't accept the PCA9685 driver and dasur h can't find it. Because at the beginning of the year it still worked until I did a Tasmota update (compiled) and screwed on the board.
2. If I change the bus address on the PCA9685, do I also have to compile the new address in Tasmota?


 ----- tested so far
1. The MCP23017 is recognized and provisional connected BME280 is also recognized.
2. I replaced the PCA9685 earlier. But the second one is not recognized; only the address 0x40 is displayed. According to my knowledge, it must be the PCA9685 or, as I misinterpreted it, the ina.

But what puzzles me is that "i2cdriver15 status" says error



Holen Sie sich Outlook für Android<https://aka.ms/ghei36>
________________________________

**Reply by sfromis** [reply]
You wrote about "I2CDriver 14 0" - which would not work. It should be "I2CDriver14 0" - that space makes a major difference to commands, but it does look like you did disable the driver.

Since your `I2CDriver` command output includes number 1, it does appear that you managed to include the driver in the build. However, since several other sensors can also appear on I2C address `0x40`, it could make sense, as barbudor suggested, to disable all the I2C drivers except for the ones you want to use.

When doing a custom build, I like to have the drivers disabled from the start, using
```
#define I2CDRIVERS_0_31   0x00000000      // 0xFFFFFFFF to Enable I2CDriver0  to I2CDriver31
#define I2CDRIVERS_32_63  0x00000000      // 0xFFFFFFFF to Enable I2CDriver32 to I2CDriver63
#define I2CDRIVERS_64_95  0x00000000      // 0xFFFFFFFF to Enable I2CDriver64 to I2CDriver95
```

Not having any experience with the PCA9685, but one potential risk factor could be if the driver has been tested on the ESP32-C2 MCU variant you use.

If you decide to switch to another I2C address for the PCA9685, instead of `0x40`, you'll indeed have to do a new build with the selected address.

Your log does show a problem, Tasmota is not able to connect to a NTP server to get correct time, which is needed for normal operation. When it comes to the MCP23017, I'd suggest you to use the enhanced mode 2, unless you specifically want the older mode 1.

**Reply by barbudor** [reply]
INA219 is i2cdriver 14 while PCA9685 is i2cdriver 1
At boot, enabled drivers are tested in order so PCA9685 driver is the 1st to kick in
If the PCA driver doesn't detect the PCA chip and move's on to the INA, it is because the chip at address 0x40 failed the test

Have you tried disabling all drivers but the PCA one ?

> I marked the INA with # in visual studio but it still showed it to me in the Gui.
????

Point 2: Yes, you have to recompile and define `USE_PCA9685_ADDR` to the new address in your `user_config_override.h`

**Comment by Basti149** [comment]
@sfromis thank you very much for the help. I actually only want to use the outputs of the MCP23017. Would it work with Mode 2?
I have to set that up in the Tasmota Gui, right?



Is there a command where I can deactivate all unnecessary drivers in one line. Otherwise it will restart every time it is deactivated. I also suspect that he is interpreting the PCA9685 as a different sensor for whatever reason (as I said this morning as an INA sensor). And as a result the driver15 commands don't work. You can tell that the PCA9685 works hardware-wise and is accessible from the 0x40 address that is displayed to me at i2csan.
But communication doesn't work for whatever reason. Tomorrow I will compile a new Tasmota32c2 version with a different bus address.
Which one would be okay?, 0x47?
________________________________

**Reply by sfromis** [reply]
One reason for me posting how you can compile to have default I2CDriver setup be all disabled is that there is no command to do them all in one go. While you *could* do many commands on one line with `Backlog0`, it is still a lot to do each and every one.

What you get with MCP23017 driver mode 2 is that you can configure pins, and then use as "normal" button/switch/relay/output, and this is done with a bit of JSON in a file, not in the web UI. This  makes usage of the expander pins much more straightforward.

And barbudor is right about the PCA9686 driver being first, meaning that it is not "highly likely" that it is another driver getting in the way. The INA219 driver "should" only be able to kick in after the PCA9685 driver did not detect the device. This is why I *speculated* about the risk of the driver not being "well-tested" with ESP32-C2.

I have no special recommendation for what to pick if you want to try with another I2C address for the PCA9685, `0x47` is probably ok to try with.

**Comment by Basti149** [comment]
No, I'll do that tomorrow to deactivate all unneeded drivers.

And to recompile it with a different address, perhaps with 0x47 (or should I use a different one). I think despite deactivating driver14 it does not release the address for the PCA9685. But it looks like he at least recognizes the chip, even if it's wrong. I'm going to bed for today. I tried a lot today but nothing worked.

But I won't give up hope with the great support from you. 🙂

________________________________

**Comment by Basti149** [comment]
thanks also for the NTP time, I will change it tomorrow.
________________________________

**Comment by Basti149** [comment]
Good evening, 
a strange miracle, I wanted to test the whole thing again this afternoon. And suddenly only two addresses 0x20 (MCP23017), 0x40 (PCA9685) were recognized, no more phantom addresses, nothing. And as you can see in the picture, I can switch the PCA9685 on/off/PWM via the console.
I honestly don't know why it suddenly works today until now (I hope forever). I have the same hardware ESP32C2, PCA9685, MCP23017 in conjunction with Tasmota v14.3.0.

Now I just have to implement the tip with Mode2 on the MCP23017. And increase the frequency so that the LEDs no longer flicker when dimming. Do I have to enter this in the console or when compiling?

Thank you both 
PS The time thing now also works because Tasmota couldn't find a DNS server.



17:51:01.978 RSL: RESULT = {"I2CScan":"Device(s) found at 0x20 0x40"}

17:52:01.006 RSL: STATE = {"Time":"2024-11-04T17:52:00","Uptime":"0T01:10:09","UptimeSec":4209,"Heap":64,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":20,"MqttCount":0,"Berry":{"HeapUsed":4,"Objects":53},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Dimmer":40,"Color":"296600","HSBColor":"96,100,40","Channel":[16,40,0],"Scheme":0,"Width":1,"Fade":"OFF","Speed":1,"LedTable":"ON","Wifi":{"AP":1,"SSId":"SmartHxxxxxxx","BSSId":"7A:8C:B5:AE:4D:10","Channel":7,"Mode":"HT40","RSSI":94,"Signal":-53,"LinkCount":1,"Downtime":"0T00:00:04"}}

17:52:01.032 RSL: SENSOR = {"Time":"2024-11-04T17:52:01","DS18B20":{"Id":"3C01B556A4AC","Temperature":24.7},"MCP230XX":{"D0":1,"D1":1,"D2":1,"D3":1,"D4":1,"D5":1,"D6":1,"D7":1,"D8":1,"D9":1,"D10":1,"D11":1,"D12":1,"D13":1,"D14":1,"D15":1,"OUT":{"OUT_D0":"OFF","OUT_D1":"OFF","OUT_D2":"OFF","OUT_D3":"OFF","OUT_D4":"OFF","OUT_D5":"OFF","OUT_D6":"OFF","OUT_D7":"OFF","OUT_D8":"OFF","OUT_D9":"OFF","OUT_D10":"OFF","OUT_D11":"OFF","OUT_D12":"OFF","OUT_D13":"OFF","OUT_D14":"OFF","OUT_D15":"OFF"}},"ESP32":{"Temperature":22.0},"TempUnit":"C"}

17:52:01.042 RSL: SENSOR = {"Time":"2024-11-04T17:52:01","PCA9685":{"PWM_FREQ":50,"INVERT":0,"PWM0":0,"PWM1":0,"PWM2":0,"PWM3":0,"PWM4":0,"PWM5":0,"PWM6":0,"PWM7":0,"PWM8":0,"PWM9":0,"PWM10":0,"PWM11":0,"PWM12":0,"PWM13":0,"PWM14":0,"PWM15":0,"END":1}}

![grafik](https://github.com/user-attachments/assets/ba5ec958-3ce6-4f76-b018-5b86ada70ff5)

**Reply by sfromis** [reply]
https://tasmota.github.io/docs/PCA9685/#driver-usage includes PWM frequency, instead of the *default* you can establish at build time.

**Reply by Basti149** [reply]
Thanks, it should work via console

**Comment by Basti149** [comment]
thank you
________________________________
Von: sfromis ***@***.***>
Gesendet: Montag, 4. November 2024 18:16
An: arendst/Tasmota ***@***.***>
Cc: Smarthome ***@***.***>; Author ***@***.***>
Betreff: Re: [arendst/Tasmota] Problems with PCA9685 (Discussion #22422)


https://tasmota.github.io/docs/PCA9685/#driver-usage includes PWM frequency, instead of the default you can establish at build time.
